### PR TITLE
[GEP 1651]: Routability Conformance

### PR DESCRIPTION
**Pull request by dprotaso**
This PR depends on the following PRs merging

- https://github.com/kubernetes-sigs/gateway-api/pull/2157
- https://github.com/kubernetes-sigs/gateway-api/pull/2144

View just the commit with the conformance test here - https://github.com/kubernetes-sigs/gateway-api/pull/2171/commits/4298344aa7a2f4659a3c310b4c7ab43c15c2972b

**What type of PR is this?**
/kind feature
/kind gep
/kind test
/area conformance

**What this PR does / why we need it**:
This adds conformance tests for the Routability feature

**Which issue(s) this PR fixes**:

Part of https://github.com/kubernetes-sigs/gateway-api/issues/1651

**Does this PR introduce a user-facing change?**:

```release-note
Add conformance tests for Gateway Routability
```

### CODE REVIEW

File: conformance/tests/gateway-routablity-broken.yaml
Comment (dprotaso) [inline conformance/tests/gateway-routablity-broken.yaml:1]: typo in filename - gonna wait for other feedback before addresssing

File: apis/v1beta1/gateway_types.go
Comment (candita) [inline apis/v1beta1/gateway_types.go:186]: The custom value feels unnecessary.  Part of the appeal of the API is portability between implementations, and allowing a custom value from the beginning blocks that portability.

File: apis/v1beta1/gateway_types.go
Comment (candita) [inline apis/v1beta1/gateway_types.go:212]: Any reason why there are brackets around `cluster's network`?

File: apis/v1beta1/gateway_types.go
Comment (candita) [inline apis/v1beta1/gateway_types.go:550]: How about enumerating a few types?  IPv4Address, IPv6Address, DNSName, etc.

File: apis/v1beta1/gateway_types.go
Comment (dprotaso) [reply apis/v1beta1/gateway_types.go:186]: This PR are about conformance tests for routability  - the types changes are in the PR here: https://github.com/kubernetes-sigs/gateway-api/pull/2157

Comment (dprotaso) [reply apis/v1beta1/gateway_types.go:186]: It was built up this way so it could be easier to review things in separate pieces but not merge things in via an intermediary stage.

File: apis/v1beta1/gateway_types.go
Comment (dprotaso) [reply apis/v1beta1/gateway_types.go:550]: see: https://github.com/kubernetes-sigs/gateway-api/pull/2171#discussion_r1259073811

File: apis/v1beta1/gateway_types.go
Comment (candita) [inline apis/v1beta1/gateway_types.go:586]:
```suggestion
	// have a routability of `Cluster` but MUST NOT include `Public`.
```

File: apis/v1beta1/gateway_types.go
Comment (candita) [inline apis/v1beta1/gateway_types.go:634]: Better to add this to the list of other conditions below.

File: apis/v1beta1/gateway_types.go
Comment (candita) [inline apis/v1beta1/gateway_types.go:590]: Isn't this considered a breaking change?  Why not add a new field for `RoutableAddresses`?

File: apis/v1beta1/gatewayclass_types.go
Comment (candita) [inline apis/v1beta1/gatewayclass_types.go:190]: nit
```suggestion
	// Implementations MAY provide a pre-defined set of GatewayClasses that
```

File: apis/v1beta1/gatewayclass_types.go
Comment (candita) [inline apis/v1beta1/gatewayclass_types.go:188]: This feels error prone, you'll need to make sure the first entry is never overwritten. Why not add a defaultRoutability field to a new type that contains the default and []GatewayRoutability?

File: conformance/utils/suite/features.go
Comment (candita) [inline conformance/utils/suite/features.go:88]:
```suggestion
	// This option indicates the GatewayClass supports routability
```

File: apis/v1beta1/gateway_types.go
Comment (candita) [inline apis/v1beta1/gateway_types.go:183]: It's helpful for me to think of this as
- Public: External
- Private: Multi-cluster
- Cluster: Internal

**Comment by candita** [comment]
I added a few comments.  Looks good, almost like a first step towards support for load balancing.

**Review by shaneutt — APPROVED** [review]
This seems pretty comprehensive, thanks @dprotaso!

Once Candace's comments are resolved, I think this looks good to me

/approve

But @robscott I think you should give this a pass too if you have some time, for the final LGTM.

**Review by shaneutt — APPROVED** [review]
This seems pretty comprehensive, thanks @dprotaso!

Once Candace's comments are resolved, I think this looks good to me

/approve

But @robscott I think you should give this a pass too if you have some time, for the final LGTM.

File: apis/v1beta1/gateway_types.go
Comment (dprotaso) [reply apis/v1beta1/gateway_types.go:550]: This is a kubebuilder directive for what the default value should be. This gets picked up when generating the CRD definition and the API server will use this value to default during resource creation when it's not set by the user.

An enumeration doesn't apply here.

File: apis/v1beta1/gateway_types.go
Comment (dprotaso) [reply apis/v1beta1/gateway_types.go:590]: The guidance I had is that go types do not need to be backwards compatible as long as the yaml format is.

File: apis/v1beta1/gateway_types.go
Comment (dprotaso) [reply apis/v1beta1/gateway_types.go:634]: Done in https://github.com/kubernetes-sigs/gateway-api/pull/2157

File: apis/v1beta1/gateway_types.go
Comment (dprotaso) [reply apis/v1beta1/gateway_types.go:586]: Fixed in https://github.com/kubernetes-sigs/gateway-api/pull/2157

File: apis/v1beta1/gateway_types.go
Comment (dprotaso) [reply apis/v1beta1/gateway_types.go:183]: Private losely maps to some internal network (eg. VPN) - so multi-cluster isn't accurate.

See the discussion on the origin GEP PR - https://github.com/kubernetes-sigs/gateway-api/pull/1653

File: apis/v1beta1/gatewayclass_types.go
Comment (dprotaso) [reply apis/v1beta1/gatewayclass_types.go:188]: This is in the status block of the GatewayClass which is reconciled by implementations. What is error prone here?

> Why not add a defaultRoutability field to a new type that contains the default and []GatewayRoutability?

This seems more complex for clients to consume - how do I get the entire list of routoabilities? do I need to concat the default with the list? then do I need to dedupe entries the list?

Comment (dprotaso) [reply apis/v1beta1/gatewayclass_types.go:188]: This comment is applicable to the PR here: https://github.com/kubernetes-sigs/gateway-api/pull/2157

If this is still a concern please comment there

**Review by howardjohn — COMMENTED** [review]
:+1:

**Comment by robscott** [comment]
Thanks @dprotaso! This will need to be rebased once #2144 and #2157 merge, but everything LGTM.

**Comment by dprotaso** [comment]
@robscott rebased

**Comment by robscott** [comment]
Thanks @dprotaso!

/lgtm

**Pull request merged**
